include/dfrn.php

```diff
@@ -0,0 +1,795 @@
+		}
+		else {
+			$type = 'text';
+		}
+
+		$entry = dfrn_entry($doc, $type, $item, $owner, true);
+		$root->appendChild($entry);
+
+	}
+
+	$atom = trim($doc->saveXML());
+
+	call_hooks('atom_feed_end', $atom);
+
+	return $atom;
+}
+
+/**
+ * @brief Create XML text for DFRN mails
+ *
+ * @param array $item message elements
+ * @param array $owner Owner record
+ *
+ * @return string DFRN mail
+ */
+function dfrn_mail($item, $owner) {
+	$doc = new DOMDocument('1.0', 'utf-8');
+	$doc->formatOutput = true;
+
+	$root = dfrn_add_header($doc, $owner, "dfrn:owner", "", false);
+
+	$mail = $doc->createElement("dfrn:mail");
+	$sender = $doc->createElement("dfrn:sender");
+
+	xml_add_element($doc, $sender, "dfrn:name", $owner['name']);
+	xml_add_element($doc, $sender, "dfrn:uri", $owner['url']);
+	xml_add_element($doc, $sender, "dfrn:avatar", $owner['thumb']);
+
+	$mail->appendChild($sender);
+
+	xml_add_element($doc, $mail, "dfrn:id", $item['uri']);
+	xml_add_element($doc, $mail, "dfrn:in-reply-to", $item['parent-uri']);
+	xml_add_element($doc, $mail, "dfrn:sentdate", datetime_convert('UTC', 'UTC', $item['created'] . '+00:00' , ATOM_TIME));
+	xml_add_element($doc, $mail, "dfrn:subject", $item['title']);
+	xml_add_element($doc, $mail, "dfrn:content", $item['body']);
+
+	$root->appendChild($mail);
+
+	return(trim($doc->saveXML()));
+}
+
+/**
+ * @brief Create XML text for DFRN friend suggestions
+ *
+ * @param array $item suggestion elements
+ * @param array $owner Owner record
+ *
+ * @return string DFRN suggestions
+ */
+function dfrn_fsuggest($item, $owner) {
+	$doc = new DOMDocument('1.0', 'utf-8');
+	$doc->formatOutput = true;
+
+	$root = dfrn_add_header($doc, $owner, "dfrn:owner", "", false);
+
+	$suggest = $doc->createElement("dfrn:suggest");
+
+	xml_add_element($doc, $suggest, "dfrn:url", $item['url']);
+	xml_add_element($doc, $suggest, "dfrn:name", $item['name']);
+	xml_add_element($doc, $suggest, "dfrn:photo", $item['photo']);
+	xml_add_element($doc, $suggest, "dfrn:request", $item['request']);
+	xml_add_element($doc, $suggest, "dfrn:note", $item['note']);
+
+	$root->appendChild($suggest);
+
+	return(trim($doc->saveXML()));
+}
+
+/**
+ * @brief Create XML text for DFRN relocations
+ *
+ * @param array $owner Owner record
+ * @param int $uid User ID
+ *
+ * @return string DFRN relocations
+ */
+function dfrn_relocate($owner, $uid) {
+
+	$a = get_app();
+
+	/* get site pubkey. this could be a new installation with no site keys*/
+	$pubkey = get_config('system','site_pubkey');
+	if(! $pubkey) {
+		$res = new_keypair(1024);
+		set_config('system','site_prvkey', $res['prvkey']);
+		set_config('system','site_pubkey', $res['pubkey']);
+	}
+
+	$rp = q("SELECT `resource-id` , `scale`, type FROM `photo`
+			WHERE `profile` = 1 AND `uid` = %d ORDER BY scale;", $uid);
+	$photos = array();
+	$ext = Photo::supportedTypes();
+	foreach($rp as $p){
+		$photos[$p['scale']] = $a->get_baseurl().'/photo/'.$p['resource-id'].'-'.$p['scale'].'.'.$ext[$p['type']];
```

Comment:
you can use `z_root()` instead of `$a-get_baseurl()` so you wouldn't need to load the whole app `$a = get_app()`